client/windows/GUIClasses.cpp

```diff
@@ -787,9 +787,95 @@ void CTavernWindow::HeroPortrait::hover(bool on)
 		GH.statusbar->clear();
 }
 
+static const std::string HD_MOD_PREFIX = "HDMOD";
+static const std::string HD_EXCHANGE_BG = HD_MOD_PREFIX + "/TRADEHD";
+
+static bool isHdLayoutAvailable()
+{
+	return CResourceHandler::get()->existsResource(ResourceID(std::string("SPRITES/") + HD_EXCHANGE_BG, EResType::IMAGE));
+}
+
+CExchangeController::CExchangeController(ObjectInstanceID hero1, ObjectInstanceID hero2)
+	:left(LOCPLINT->cb->getHero(hero1)), right(LOCPLINT->cb->getHero(hero2))
+{
+}
+
+std::function<void()> CExchangeController::onMoveArmyToLeft()
+{
+	return [&]() { hdModQuickExchangeArmy(false); };
+}
+
+std::function<void()> CExchangeController::onMoveArmyToRight()
+{
+	return [&]() { hdModQuickExchangeArmy(true); };
+}
+
+std::function<void()> CExchangeController::onSwapArmy()
+{
+	return [&]() { 
+		auto cb = LOCPLINT->cb;
+
+		for(SlotID i = SlotID(0); i.validSlot(); i.advance(1))
+		{
+			if(left->hasStackAtSlot(i) || right->hasStackAtSlot(i))
+				cb->swapCreatures(left, right, SlotID(i), SlotID(i));
+		}
+	};
+}
+
+void CExchangeController::hdModQuickExchangeArmy(bool leftToRight)
+{
+	auto source = leftToRight ? left : right;
+	auto target = leftToRight ? right : left;
+	auto cb = LOCPLINT->cb;
+
+	boost::thread([=]
+	{

```

Comment:
Please explain how threading helps here (but not in function above, both do server call for each slot).

Reply:
cb->waitTilRelaize locks main thread. For easier implementation here I need to see the results of cb invokations so I need to use waitTilRealize and thus I need a separate thread.

Reply:
The question is why is a new thread here but in `CExchangeController::onSwapArmy()` for same amount of `cb` calls is not? Also if you spawn a thread you need (at least shared) lock of gs there, but this in turn will still lock UI. And with acquired lock you could use unlockGsWhenWaiting.

Reply:
Because I do not need the result of those calls in swapArmy

Reply:
Btw I think you spotted a nice bug if the first stack for swapping is the last one and the same slot of another army is empty

Reply:
Is it ok now?

Reply:
Yes it is